components/common/src/command/package/install.rs

```diff
@@ -502,43 +625,139 @@ impl<'a> InstallTask<'a> {
 
     /// Adapter function to retrieve an installed package given an
     /// identifier, if it exists.
-    fn installed_package(&self, ident: &PackageIdent) -> Option<PackageInstall> {
-        PackageInstall::load(ident, Some(self.fs_root_path)).ok()
+    fn installed_package(&self, ident: &FullyQualifiedPackageIdent) -> Option<PackageInstall> {
+        PackageInstall::load(ident.as_ref(), Some(self.fs_root_path)).ok()
+    }
+
+    /// Checks for the latest installed package or cached artifact that matches a given package
+    /// identifier and returns a fully qualified package identifer if a match exists.
+    fn latest_installed_or_cached(
+        &self,
+        ident: &PackageIdent,
+    ) -> Result<FullyQualifiedPackageIdent> {
+        let latest_installed = self.latest_installed_ident(&ident);
+        let latest_cached = self.latest_cached_ident(&ident);
+        debug!(
+                "latest installed: {:?}, latest_cached: {:?}",
+                &latest_installed,
+                &latest_cached,
+            );
+        let latest = match (latest_installed, latest_cached) {
+            (Ok(pkg_install), Err(_)) => pkg_install,
+            (Err(_), Ok(pkg_artifact)) => pkg_artifact,
+            (Ok(pkg_install), Ok(pkg_artifact)) => {
+                if pkg_install.as_ref() > pkg_artifact.as_ref() {
+                    pkg_install
+                } else {
+                    pkg_artifact
+                }
+            }
+            (Err(_), Err(_)) => return Err(Error::PackageNotFound),
+        };
+        debug!("offline mode: winner: {:?}", &latest);
+
+        Ok(latest)
     }
 
-    // TODO (CM): This could return a plain bool IF we could ensure
-    // above that the package identifier is FULLY QUALIFIED
-    fn is_artifact_cached(&self, ident: &PackageIdent) -> Result<bool> {
-        Ok(self.cached_artifact_path(ident)?.is_file())
+    fn latest_installed_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        match PackageInstall::load(ident, Some(self.fs_root_path)) {
+            Ok(pi) => FullyQualifiedPackageIdent::from(pi.ident().clone()),
+            Err(_) => Err(Error::PackageNotFound),
+        }
+    }
+
+    fn latest_cached_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        let filename_glob = {
+            let mut ident = ident.clone();
+            if ident.version.is_none() {
+                ident.version = Some(String::from("_VERSION_"));
+            }
+            if ident.release.is_none() {
+                ident.release = Some(String::from("_RELEASE_"));
+            }
+            match ident.archive_name() {
+                Some(name) => name.replace("_VERSION_", "?*").replace("_RELEASE_", "?*"),
+                None => return Err(Error::PackageNotFound),
+            }
+        };
+        let glob_path = self.artifact_cache_path.join(filename_glob);
+        let glob_path = glob_path.to_string_lossy();
```

Comment:
There definitely shouldn't be any non-unicode sequences here, right? Would
```rust
        let glob_path = glob_path.to_str().unrwap();
```
be safer and obviate the need for the `as_ref()`s later?

Reply:
There's a super low chance of hitting any non-Unicode chars, correct. I usually use the [`PathBuf.to_string_lossy()`](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.to_string_lossy) in these situations as the result is a `Cow<str>` which delays any further allocations until necessary. If we hit a non-Unicode character, it's true that this code would add those strange `�` characters in the resulting string, but we also wouldn't thread panic and in the worst case not find a cached artifact. Since we don't allow non-Unicode characters in the package identifiers (in the `PackageInstall` `FromStr` impl), we're left with the filesystem path having one of these characters. This is one place where I'd rather let this code work through rather than panicking and having to cope with a source location and an unknown unwrap call.